### PR TITLE
Add data export functionality

### PR DESCRIPTION
**Pull request by haticekaratay**
### Description

This pull request adds data export functionality to the export plugin.

Fixes #<Issue Number>

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:61]: this already inherits from `DatasetMultiSelectMixin` so should not need to add anything new

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:142]: we should be able to just observe `dataset_selected` and _assume_ that the data remains unchanged once selected (unlike subsets which can change while they're selected).

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:281]: should have similar logic as tables/subsets to check if the extension is already provided and only add it otherwise

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:283]: does this overwrite by default?

File: jdaviz/configs/default/plugins/export/export.vue
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.vue:111]: instead of closing the second div, I think the `<div class="section-description">` should just be removed (seems to erroneously have been introduced in #2760)

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:280]: can we check for `data_invalid_msg` and raise the error if its set?

**Review by kecnry — COMMENTED** [review]
I know its still in draft - but just noticed a few things on a quick skim and didn't want to forget!

File: jdaviz/configs/default/plugins/export/export.vue
Comment (kecnry) [reply jdaviz/configs/default/plugins/export/export.vue:111]: I fixed this in #2772 (so that we can have the fix in main before the PR is ready for merge).  Once that's merged, rabasing on top of main and dealing with the conflict should fix the problem.

File: jdaviz/configs/default/plugins/export/export.py
Comment (haticekaratay) [reply jdaviz/configs/default/plugins/export/export.py:281]: Thanks, Kyle, for the feedback. Somehow, I thought we should only write data in a FITS file for now. Should I add a drop-down section for these formats [ASCII and FITS]?

Comment (kecnry) [reply jdaviz/configs/default/plugins/export/export.py:281]: only fits is ok, but then we should still check if filename already ends with fits

**Review by cshanahan1 — COMMENTED** [review]
Im testing this out (on the imviz example notebook). When I select a dataset to export, it produces a message that 'file was exported to imviz_export.fits', however the file doesn't exist. I would expect it to be output in the cwd. Can you test if this works for you?

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:291]: Let's make sure we don't raise a success snackbar message in cases where `write` is not called (although these cases - like images that @cshanahan1 brought up - should either be supported OR included in the `data_invalid_msg` logic).

```suggestion
            if isinstance(self.dataset.selected_obj, Spectrum1D):
                self.dataset.selected_obj.write(Path(filename), overwrite=True)
            else:
                raise NotImplementedError(f"exporting {self.dataset.selected} not supported")

```

File: jdaviz/configs/default/plugins/export/export.py
Comment (cshanahan1) [inline jdaviz/configs/default/plugins/export/export.py:226]: is there something about this specific data unit that makes it unable to save? just curious what the back story on this is since ive never run into it

File: jdaviz/configs/default/plugins/export/export.py
Comment (haticekaratay) [reply jdaviz/configs/default/plugins/export/export.py:153]: I see the docs are updated on @cshanahan1's export plugin plots PR. Thanks!

**Comment by haticekaratay** [comment]
> Im testing this out (on the imviz example notebook). When I select a dataset to export, it produces a message that 'file was exported to imviz_export.fits', however the file doesn't exist. I would expect it to be output in the cwd. Can you test if this works for you?



> Im testing this out (on the imviz example notebook). When I select a dataset to export, it produces a message that 'file was exported to imviz_export.fits', however the file doesn't exist. I would expect it to be output in the cwd. Can you test if this works for you?

This issue has been resolved thanks to Kyle's suggestion. Now, when there is an error, a snack bar message is displayed and export is prevented.

File: jdaviz/configs/default/plugins/export/export.py
Comment (cshanahan1) [reply jdaviz/configs/default/plugins/export/export.py:226]: ahhh i didn't know this! do we want to check if other units produce this warning? is it any sort of combination of units that does this?

**Comment by kecnry** [comment]
If we don't support exporting images yet, then we should detect that in advance and set the `data_invalid_msg`.  The `NotImplementedError`/snackbar should just be a backup mechanism to catch cases that may not be caught in advance.

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:231]:
```suggestion
            if isinstance(self.dataset.selected_obj, NDData):
                self.data_invalid_msg = f"Export isn't implemented for the NDdata objects."
            elif not isinstance(self.dataset.selected_obj, Spectrum1D):
                self.data_invalid_msg = "Export is not implemented for this type of data"
            else:

```

File: jdaviz/configs/default/plugins/export/export.py
Comment (haticekaratay) [reply jdaviz/configs/default/plugins/export/export.py:227]: Not entirely certain as I have only tested against the existing notebooks. Do you have any suggestions on which units I should look out for?

Comment (pllim) [reply jdaviz/configs/default/plugins/export/export.py:227]: This is probably too specific to catch all the edge cases (and we have a lot that snuck into MAST over the years). The cause is that somewhere in the chain, something is doing `this_unit.to_string(format="fits")`. It works in generic format, but not as part of the FITS standard. Do you still have the full traceback? Where exactly is this happening?

Comment (kecnry) [reply jdaviz/configs/default/plugins/export/export.py:227]: Maybe we can just try/except the string formatting in advance then?

**Comment by cshanahan1** [comment]
What we want here is to just enable exporting of data that was imported? Not to save, for example, the 2d image arrays and Spectrum1D of a cube slice that is currently being shown?

File: jdaviz/configs/default/plugins/export/export.py
Comment (pllim) [reply jdaviz/configs/default/plugins/export/export.py:227]: What exactly is the use case that triggers this? "DN" (per second or no) usually is for uncalibrated data. If I understand the context more, then I can hopefully give better options.

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/export/export.py:230]: if Spectrum1D inherits from NDData, then this won't work.  My suggestion would be to mirror the same logic used when actually exporting.

**Comment by cshanahan1** [comment]
I'm getting the same 'can't export NDData' no matter what item i have selected for export.

An input Spectrum1D
Screen Shot 2024-04-01 at 4 06 54 PM

And a spectrum loaded from a file:
Screen Shot 2024-04-01 at 4 08 44 PM

File: jdaviz/configs/default/plugins/export/export.py
Comment (cshanahan1) [inline jdaviz/configs/default/plugins/export/export.py:227]: If the data doesn't come from a plugin, it can't be exported (yet). If it does come from a plugin, it must either be a Spectrum1D or CCDData. Is this the output format of all plugin data currently, would this ever be encountered? (cc @kecnry).

**Review by cshanahan1 — COMMENTED** [review]
Left a few comments - since the scope of this seems to be to only export plugin generated data, it works for that. I also see it correctly display the warning message when non-generated data is selected.

A few things that I think would be important to address in a follow up:

1. To get the moment map to export, the export plugin can only find it if it's been added to the viewer. I personally think it should be exportable whenever generated, even if it wasn't added to a viewer. I think this same issue would apply to other plugin generated data as well.
2. Related to 1, you CAN save the moment map without adding it to the viewer if you 'save to fits' in the moment map plugin. I think this save to fits button should be removed and we should just link to the export plugin with a button there.

(I'll approve once tests pass)

File: jdaviz/configs/default/plugins/export/tests/test_export.py
Comment (cshanahan1) [inline jdaviz/configs/default/plugins/export/tests/test_export.py:191]: can you check that the data_invalid_msg is set correctly in both of these tests?

**Comment by kecnry** [comment]
> To get the moment map to export, the export plugin can only find it if it's been added to the viewer. I personally think it should be exportable whenever generated, even if it wasn't added to a viewer. I think this same issue would apply to other plugin generated data as well.

I think this is a more general bug with the `DatasetSelect` component (that will definitely need to be fixed, but might be out of scope here).  [:cat:](https://jira.stsci.edu/browse/JDAT-4383).

File: jdaviz/configs/default/plugins/export/export.py
Comment (pllim) [reply jdaviz/configs/default/plugins/export/export.py]: Depending on where the offending code is, it needs to be more permissive, maybe like this:

```python
>>> u.Unit("DN/s", format="fits", parse_strict="silent").to_string()
'DN/s'
```

Or we can code in exception rule like you have, though one rule per every possible string out there can come back to bite us.

**Comment by cshanahan1** [comment]
Yeah I don't think it needs to be done in this PR, but as a follow up. The export functionality in the moment map plugin and in the export plugin for the same thing, but one requiring it in the viewer and one not is just confusing.

I would say this is ready to merge after resolving conflicts?

**Review by cshanahan1 — APPROVED** [review]

File: jdaviz/configs/default/plugins/export/export.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/export/export.py]: We decided to leave this check out and let the snackbar catch this case.  @haticekaratay will create a follow-up ticket to track down and fix the cause.  Thanks all!

**Pull request merged**
